File: Java/src/ospf/ospf.java
package ospf;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.Scanner;

public class ospf {
    public static void main(String[] args) throws FileNotFoundException {
        File f = new File("src/ospf/test.txt");
        ArrayList<Node> list = initNodeList(f);
        for (int i = 0; i < list.size(); i++) {
            createRoutineTable(list, list.get(i));
            ArrayList<TC> q = list.get(i).getConfirmed();
            System.out.printf("-------confirmed of node %c-------\n", list.get(i).getNumber() + 64);
            for (TC tc : q) {
                System.out.printf("(%c,%d,", tc.getDestination() + 64, tc.getCost());
                if (tc.getNextHop() == 0) {
                    System.out.println("-)");
                } else {
                    System.out.printf("%c)\n", tc.getNextHop() + 64);
                }
            }
        }
    }

    public static ArrayList<LSP> locateLSPNode(ArrayList<Node> list, int number) {
        int i = 0;
        for (; i < list.size(); i++) {
            if (list.get(i).getNumber() == number) {
                return list.get(i).getLSPList();
            }
        }
        return null;
    }

    public static void createRoutineTable(ArrayList<Node> list, Node q) {
        q.getConfirmed().add(new TC(q.getNumber(), 0, 0));
        TC cp = q.getConfirmed().get(q.getConfirmed().size() - 1);
        do {
            ArrayList<LSP> r = locateLSPNode(list, cp.getDestination());
            assert r != null;
            for (LSP lsp : r) {
                int newCost = cp.getCost() + lsp.getCost();
                TC s = ifExist(lsp.getNeighbor(), q.getTentative());
                if (!(ifExist(lsp.getNeighbor(), q.getConfirmed()) != null || s != null)) {
                    q.getTentative().add(new TC(lsp.getNeighbor(), newCost, cp.getNextHop() == 0 ? lsp.getNeighbor() : cp.getNextHop()));
                } else if (s != null && newCost < s.getCost()) {
                    s.setCost(newCost);
                    s.setNextHop(cp.getNextHop() == 0 ? lsp.getNeighbor() : cp.getNextHop());
                }
            }
            if (q.getTentative().isEmpty()) {
                break;
            } else {
                moveTCNde(q.getConfirmed(), q.getTentative());
                cp = q.getConfirmed().get(q.getConfirmed().size() - 1);
            }
        } while (true);
    }

    public static void moveTCNde(ArrayList<TC> confirmed, ArrayList<TC> tentative) {
        TC minCost = tentative.get(0);
        for (int i = 1; i < tentative.size(); i++) {
            if (tentative.get(i).getCost() < minCost.getCost()) {
                minCost = tentative.get(i);
            }
        }
        confirmed.add(new TC(minCost.getDestination(), minCost.getCost(), minCost.getNextHop()));
        tentative.remove(minCost);
    }

    public static TC ifExist(int number, ArrayList<TC> list) {
        int i = 0;
        for (; i < list.size(); i++) {
            if (list.get(i).getDestination() == number) {
                return list.get(i);
            }
        }
        assert i != list.size();
        return null;//unmeaning
    }

    public static ArrayList<Node> initNodeList(File f) throws FileNotFoundException {
        int number, neighbor, cost;
        ArrayList<Node> list = new ArrayList<>();
        int current = 0;
        Scanner sc = new Scanner(f);
        while (sc.hasNext()) {
            number = sc.nextInt();
            neighbor = sc.nextInt();
            cost = sc.nextInt();
            if (current != number) {
                current = number;
                Node q = new Node();
                q.setNumber(number);

                q.setLSPList(new ArrayList<>());
                q.getLSPList().add(new LSP(neighbor, cost));

                q.setTentative(new ArrayList<>());

                q.setConfirmed(new ArrayList<>());
                list.add(q);
            } else {
                list.get(list.size() - 1).getLSPList().add(new LSP(neighbor, cost));
            }
        }
        return list;
    }
}

File: Python/ospf/ospf.py
import os

class LSP():
    def __init__(self, neighbor, cost):
        self.neighbor = neighbor
        self.cost = cost

class TC():
    def __init__(self, destination, cost, nextHop):
        self.destination = destination
        self.cost = cost
        self.nextHop = nextHop

class Node():
    def __init__(self, number, LSPList, tentative, confirmed):
        self.number = number
        self.LSPList = LSPList
        self.tentative = tentative
        self.confirmed = confirmed

def initNodeList(f):
    number, neighbor, cost = 0, 0, 0
    list = []
    current = 0
    for line in f:
        number, neighbor, cost = map(int, line.strip().split())
        if current != number:
            current = number
            q = Node(number, [], [], [])
            q.LSPList.append(LSP(neighbor, cost))
            list.append(q)
        else:
            list[len(list) - 1].LSPList.append(LSP(neighbor, cost))
    return list

def locateLSPNode(list, number):
    for i in range(len(list)):
        if list[i].number == number:
            return list[i].LSPList
    return None

def ifExist(number, list):
    for i in range(len(list)):
        if list[i].destination == number:
            return list[i]
    return None

def moveTCNode(confirmed, tentative):
    minCost = tentative[0]
    for i in range(1, len(tentative), 1):
        if tentative[i].cost < minCost.cost:
            minCost = tentative[i]
    confirmed.append(TC(minCost.destination, minCost.cost, minCost.nextHop))
    tentative.remove(minCost)

def createRoutineTable(list, q):
    q.confirmed.append(TC(q.number, 0, 0))
    cp = q.confirmed[-1]
    while True:
        r = locateLSPNode(list, cp.destination)
        assert r != None
        for lsp in r:
            newCost = cp.cost + lsp.cost
            s = ifExist(lsp.neighbor, q.tentative)
            if not (ifExist(lsp.neighbor, q.confirmed) != None or s != None):
                q.tentative.append(TC(lsp.neighbor, newCost, lsp.neighbor if cp.nextHop == 0 else cp.nextHop))
            elif s != None and newCost < s.cost:
                s.cost = newCost
                s.nextHop = lsp.neighbor if cp.nextHop == 0 else cp.nextHop
        if len(q.tentative) == 0:
            break
        else:
            moveTCNode(q.confirmed, q.tentative)
            cp = q.confirmed[-1]
        
def main():
    if os.path.exists(r"Python\ospf\test.txt"):
        list = None
        with open(r"Python\ospf\test.txt", "r") as f:
            list = initNodeList(f)
        for i in range(len(list)):
            createRoutineTable(list, list[i])
            q = list[i].confirmed
            print(f"-------confirmed of node {chr(list[i].number + 64)}-------")
            for tc in q:
                print(f"({chr(tc.destination + 64)},{tc.cost},", end="")
                if tc.nextHop == 0:
                    print("-)")
                else:
                    print(f"{chr(tc.nextHop + 64)})")
    else:
        print("file not found")

if __name__ == "__main__":
    main()